rust/lance/src/encodings/binary.rs

```diff
@@ -311,23 +336,88 @@ impl<'a, T: ByteArrayType> Decoder for BinaryDecoder<'a, T> {
         let positions = self
             .get_positions(start as usize..(end + 1) as usize)
             .await?;
+        // Use indices and positions to pre-allocate an exact-size buffer
+        let capacity = indices
+            .iter()
+            .map(|i| {
+                let relative_index = (i.unwrap() - start) as usize;
+                let start = positions.value(relative_index) as usize;
+                let end = positions.value(relative_index + 1) as usize;
+                end - start
+            })
+            .sum();
+        let mut buffer = MutableBuffer::with_capacity(capacity);
+
+        let offsets_capacity = std::mem::size_of::<T::Offset>() * (indices.len() + 1);
+        let mut offsets = MutableBuffer::with_capacity(offsets_capacity);
+        let mut offset = T::Offset::from_usize(0).unwrap();
+        // Safety: We allocated appropriate capacity just above.
+        unsafe {
+            offsets.push_unchecked(offset);
+        }
+
         let chunks = plan_take_chunks(&positions, indices, MIN_IO_SIZE)?;
 
-        let arrays = stream::iter(chunks)
-            .zip(repeat_with(|| positions.clone()))
-            .map(|(indices, positions)| async move {
-                self.take_internal(positions.as_ref(), &indices).await
+        let positions_ref = positions.as_ref();
+        futures::stream::iter(chunks)
+            .map(|chunk| async move {
+                let start = chunk.indices.value(0);
+                let end = chunk.indices.value(chunk.indices.len() - 1);
+                let array = self
+                    .get_range(positions_ref, start as usize..end as usize + 1)
+                    .await?;
+                Result::Ok((chunk, start, array))
             })
             .buffered(num_cpus::get())
-            .try_collect::<Vec<_>>()
+            .try_for_each(|(chunk, start, array)| {
+                let array: &GenericByteArray<T> = array.as_bytes();
+
+                // Faster to do one large memcpy than O(n) small ones.
+                if chunk.is_contiguous {
+                    buffer.extend_from_slice(array.value_data());
+                }
+
+                // Append each value to the buffer in the correct order
+                for index in chunk.indices.values() {
+                    if !chunk.is_contiguous {
+                        let value = array.value((index - start) as usize);
+                        let value_ref: &[u8] = value.as_ref();
+                        buffer.extend_from_slice(value_ref);
+                    }
+
+                    offset += array.value_length((index - start) as usize);
```

Comment:
Does `start` need to change in each iteration of this inner loop?  E.g. if `start` is 0 and there were 3 indices in `chunk` (10, 20, 50) then wouldn't this do:

```
offset += 10
offset += 20
offset += 50
```

But wouldn't you want...
```
offset += 10
offset += 10
offset += 30
```

Reply:
`start` is sort of a bad name here. I think it should instead be called `chunk_offset`. Basically `index` is relative to the overall array, but array is just a slice of the array, with an offset of `chunk_offset`. `array.value_length()` gets the length of the `i`th value.

So the situation is more like:

```python
whole_array = ["a", "aa", "aaa", "aaaa", "aaaaa"]
start = 1
array = ["aa", "aaa"]
chunk.indices = [1, 2]
```

So to get the length of the index `1`, which corresponds to `"aa"` in the whole array, we need to get the length of `1 - start` which is `0`th element in `array`.

Reply:
Got it.  Thanks for the explanation.